### ISSUE TITLE
Enable/disable/list repos using redhat_subscription module

### ISSUE THREAD

**Issue by rvalente** [ansible/ansible#28159]
_From @ghost on March 22, 2016 11:7_

##### Issue Type:
- Feature Idea
##### Plugin Name:

redhat_subscription

##### COMPONENT NAME
lib/ansible/modules/packaging/os/redhat_subscription.py

##### Ansible Version:

N/A
##### Ansible Configuration:

N/A
##### Environment:

N/A
##### Summary:

redhat_subscription module does not contains options to enable/disable/list repos.
When connecting a server to the rhn or satellite and assign a subscription without activation keys, its necessary to enable or disable repos.
##### Steps To Reproduce:

subscription-manager command provides those options:
- Enable: subscription-manager repos --enable repo [--enable repo] [--enable regexp]
- Disable: subscription-manager repos --disable repo [--disable repo] [--disable regexpr]
- List all: subscription-manager repos --list
- List enabled: subscription-manager repos --list-enabled
- List disabled: subscription-manager repos --list-disabled

_Copied from original issue: ansible/ansible-modules-core#3300_

**Comment by rvalente** [comment]
_From @jfenal on June 8, 2016 23:27_

As for needs:
- get list of repos currently enabled on the machine
- enable if not yet enabled a repos or a list of repos (in one command, as subscription-manager is slooooow)
- same for disabling a list of repos, with special '*' to disable all repos (and build on a clean state)

**Comment by rvalente** [comment]
_From @samdoran on June 13, 2016 19:55_

I would like this functionality as well. I'm wondering about the best way to go about implementing it.

Would it make sense to add this to the existing `redhat_subscription` module given the `state:` parameter would then be ambiguous, i.e., is the _repo_ enabled ( `repos --enable` vs `repos --disable`) or is the _subscription_ enabled (`register` vs. `unregister`)? Adding a `repo_state` parameter would remove the amibiguity, but seems a little gross.

Or would it be better to split this out into a separate `redhat_repository` module so it's clear exactly what is being enabled/disabled ([singular name](https://github.com/ansible/proposals/issues/10) to make @abadger happy)?

**Comment by rvalente** [comment]
_From @minusdavid on July 22, 2016 1:2_

Since the modules "rhn_channel" and "rhn_register" already exist, I think it would be reasonable and consistent to create a new module called "redhat_repository" to go alongside "redhat_subscription". 

By the way, I'd also like this functionality. I wonder how people are managing their repos using subscription-manager in Ansible at the moment. Using the command/shell modules I'm guessing?

**Comment by rvalente** [comment]
_From @minusdavid on July 22, 2016 7:11_

As a short-term measure, I'm writing a custom "site_facts" module which will parse redhat.repo, and then I can use that with some conditionals to do "command: subscription-manager repos...".

It's not really that different from the concept behind http://docs.ansible.com/ansible/include_vars_module.html I imagine... except instead of YAML/JSON it's the INI type format that Python uses with ConfigParser.

I'm going to look at it again on Monday, and if I can make it nice and clean, I'll post it on my own Github account for anyone else who wants to use it.

**Comment by rvalente** [comment]
_From @jfenal on July 22, 2016 14:56_

I've started writing such a module, but was interrupted before I could finish a first version I could test.
Have a look at : https://github.com/jfenal/ansible-modules-jfenal
The logic is mostly there, BUT not tested at all, no JSON output, no error handling. It clearly lacks some more code to be functional, but close.

Please note my disclaimer : my first ansible module, and moreover, my first attempt to write (with lots of copy/pasting) python code. Cave canem :)

**Comment by rvalente** [comment]
_From @samdoran on July 22, 2016 19:41_

I have a [role](https://galaxy.ansible.com/samdoran/redhat-subscription/) that's using `command` to manage subscriptions and repos.

**Comment by rvalente** [comment]
_From @minusdavid on July 23, 2016 14:55_

Kudos for your work, guys!

@samdoran I take it that's not idempotent, so you use tags to just run it once during initial setup?

@jfenal Looks ambitious! I opted for a local custom module, which you can see at this [Gist](https://gist.github.com/minusdavid/2774a8d6df5ded46b1b63e05f214da01)

**Comment by rvalente** [comment]
_From @minusdavid on July 23, 2016 14:57_

Btw, it works well enough. I like the idea of a core redhat_repository module best, but this will certainly do the job for now.

**Comment by rvalente** [comment]
_From @minusdavid on July 23, 2016 15:3_

Just realized that you both appear to work at Red Hat.

I'm super curious about if/how RH is developing Ansible to help manage RHEL...

I've only been using RHEL for a couple of weeks, but I've used Ansible for a couple years. So far so good using it to setup RHEL servers. The only real sticking point so far has been this issue. Well, and this [one](https://github.com/ansible/ansible/issues/16006), but perhaps this comment section isn't the best place for discussing such things.

**Comment by rvalente** [comment]
_From @detiber on September 8, 2016 17:10_

@barnabycourt not sure if you've seen this bug

**Comment by rvalente** [comment]
_From @ansibot on October 6, 2016 14:57_

@barnabycourt, @alikins, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by rvalente** [comment]
_From @jfenal on November 28, 2016 22:43_

https://github.com/ansible/ansible-modules-extras/pull/3559

**Comment by rvalente** [comment]
_From @ansibot on December 7, 2016 18:6_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by rvalente** [comment]
_From @jfenal on December 27, 2016 17:8_

For followers of this thread, please have a look at https://github.com/jfenal/ansible-modules-jfenal, all feedbacks are welcome before I submit it to ansible core.

**Comment by rvalente** [comment]
_From @detiber on January 3, 2017 19:46_

@barnabycourt might be of interest to you.

**Comment by ansibot** [comment]
@rvalente Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by alikins** [comment]
ping @alikins, @barnabycourt, @kahowell

**Comment by giovannisciortino** [comment]
I submitted a PR for a new module redhat_subscription based on the needs described in this comment and on my sysadmin experience on subscription-manager. I'm glad to receive suggestions and review to improve it.

**Comment by jmtd** [comment]
This would be *very* useful. Currently we are doing e.g.

```
- shell: "{{ item }}"                                                                                                   
  with_items:                                                                                                           
    - subscription-manager repos --disable='*'                                                                          
    - >                                                                                                                 
      subscription-manager repos                                                                                        
      --enable="rhel-7-server-rpms"                                                                                     
      --enable="rhel-7-server-extras-rpms"                                                                              
      --enable="rhel-7-server-optional-rpms"      
```
which is obviously not ideal. I will take a look at @giovannisciortino's PR to see what I think (which is #28292 btw)

**Comment by giovannisciortino** [comment]
@jmtd I have to do apply some changes on PR #28292 described in the review of "samdoran on Dec 15, 2017". I'm thinking to remove the code related to listing subscription-manager repositories and leave only the features to enable/disable repositories. I need some spare time to work on it, I hope to have it soon maybe in the next weekend.
In the meantime if you want test the enable/disable repositories features of this PR you are welcome.
